base/intfuncs.jl

```diff
@@ -195,6 +195,11 @@ end
 ^(x::Number, p::Integer)  = power_by_squaring(x,p)
 ^(x, p::Integer)          = power_by_squaring(x,p)
 
+# x^p for any literal integer p is lowered to x^Val{p},
+# to enable compile-time optimizations specialized to p.
+# However, we still need a fallback that calls the general ^:
+^{p}(x, ::Type{Val{p}}) = x^p
```

Comment:
How about we rewrite this as `^(x, p) = internal_pow(x, p)` and then dispatch on `Val` in `internal_pow`? I ask because this is going to be a magnet for ambiguities since packages are likely to specialize on the first argument. E.g., https://github.com/ajkeller34/Unitful.jl/pull/54